### ISSUE TITLE
Help on implementation for point annotations.

### ISSUE THREAD

**Issue by alfieroddan** [CircleRadon/APro#3]
Hi I was trying to use your code to implement the weak annotation section of your paper.

```
class AproLoss(nn.Module):
    def __init__(self):
        super().__init__()
        # partial cross entropy
        self.partialCE = SparseWeightedCE()
        # apro
        self.global_apro = Global_APro()
        self.local_apro = Local_APro(kernel_size=5, zeta_s=0.15) #set kernel_size and zeta_s
        self.mst = MinimumSpanningTree(Global_APro.norm2_distance)
        # pca n_component
        self.q = 1

    def compute_pca(self, x):
        B, C, H, W = x.shape
        pca_images = torch.empty((B, 1, H, W), device=x.device)
        for b in range(B):
            A = x[b, :, :, :].flatten(1).T
            U, S, V = torch.pca_lowrank(A, q=1)
            projection = torch.matmul(A, V)
            pca_images[b, :, :, :] = projection.T.reshape(1, H, W)
        
        # normalize
        # Min-max normalization
        min_vals = torch.min(pca_images, dim=0).values
        max_vals = torch.max(pca_images, dim=0).values
        # Avoid division by zero in case max == min
        normalized_ims = (pca_images - min_vals) / (max_vals - min_vals + 1e-10)
        return normalized_ims 

    def forward(self, x, y_hat, y):
        # x: B, C, H, W
        # y_hat B, classes, H, W
        # partial cross entropy
        partial = self.partialCE(y_hat, y)
        # compute PCA
        # B, 1, H, W
        pca_imgs = self.compute_pca(x)

        # compute image tree
        img_mst_tree = self.mst(pca_imgs)

        # softmax y
        _, class_hat = torch.max(y_hat, dim=1)
        # Cls: B, H, W
        class_hat = class_hat.float()
        # y: B, H, W
        y = y.float()

        # psuedo label for global info
        soft_pseudo = self.global_apro(class_hat, pca_imgs, img_mst_tree, zeta_g=0.01)

        # labelled region only
        labelled_regions = (y != 0)

        # compute difference between psuedo labels and other
        loss_global_term = torch.abs(soft_pseudo-y) * labelled_regions
        # normalize the loss
        n_regions = labelled_regions.sum().clamp(min=1)
        loss_global = loss_global_term.sum() / n_regions

        # local term
        soft_pseudo = self.local_apro(pca_imgs, y, labelled_regions)
        loss_local_term = torch.abs(y - soft_pseudo) * labelled_regions
        loss_local_term = loss_local_term.sum() / labelled_regions.sum()
        loss_local = loss_local_term
        return partial + loss_global + loss_local
```

Is the above implementation correct?

I'm worried about a couple of things.

1. The tree / GP being inside the loss function, should they be in the model?
2. The shapes and sizes of my labels, in your example you sum over dim 1, 2, and 3. What shape is box regions?

Any help would be greatly appreciated, currently the loss looks like this.
![Screenshot 2024-03-03 at 18 51 02](https://github.com/CircleRadon/APro/assets/51797647/5b44958b-c47d-426b-a0a8-590c59ff9df4)

**Comment by CircleRadon** [comment]
Hi, @alfieroddan 
1. The tree/GP is utilized solely in the loss function, not incorporated in the model.
2. In our implementation, the shape of box_mask_target is [box_num, 1, h, w].

Upon reviewing your implementation, I found some confusion,

1. The use of PCA is unnecessary in this context. Instead, you should directly construct a minimum spanning tree from the raw image, for example, by calling self.mst(raw_image).
2. The method aims to generate the pseudo labels of  unlabeled regions rather than labeled points.  

To facilitate integration with your existing code, I have made some minor amendments and uploaded the updated code.

And I made some simple modifications to your code, you can pull the newest code and try it:

```
img_mst_tree = self.mst(raw_image)
soft_pseudo = self.global_apro(y_hat, raw_imgs, img_mst_tree, zeta_g=0.001)
soft_pseudo = self.global_apro(soft_pseudo, y_hat, img_mst_tree, zeta_g=0.05)
unlabelled_regions = (y == 0)
loss_global_term = torch.abs(soft_pseudo-y) *unlabelled_regions
n_regions = unlabelled_regions.sum().clamp(min=1)
loss_global = loss_global_term.sum() / n_regions


soft_pseudo = self.local_apro(raw_imgs, y)
loss_local_term = torch.abs(y - soft_pseudo) * unlabelled_regions
loss_local_term = loss_local_term.sum() / unlabelled_regions.sum()
loss_local = loss_local_term
return partial + loss_global + loss_local
```

The main codes are shown above, and there may be some minor issues that you need to adjust for your project.

**Comment by alfieroddan** [comment]
Thanks for getting back to me so quickly,

Sorry I should've mentioned that PCA is done as my image has more then 3 channels.

This great, thank you very much. I will implement the changes.

By the way I had to modify the GP_cuda implementation to compile for Pytorch > 1.10. If you would like the changes let me know!

**Comment by alfieroddan** [comment]
```python
class AproLoss(nn.Module):
    def __init__(self):
        super().__init__()
        # partial cross entropy
        self.partialCE = SparseWeightedCE()
        # apro
        self.global_apro = Global_APro()
        self.local_apro = Local_APro(kernel_size=5, zeta_s=0.15) #set kernel_size and zeta_s
        self.mst = MinimumSpanningTree(Global_APro.norm2_distance)
        # pca n_component
        self.q = 1

    def compute_pca(self, x):
        B, C, H, W = x.shape
        # Reshape data for PCA (combine all dimensions except for the channels)
        data_reshaped = x.view(-1, C)
        
        # Mean centering
        mean_centered_data = data_reshaped - data_reshaped.mean(dim=0)
        
        # Compute covariance matrix
        cov_matrix = torch.mm(mean_centered_data.t(), mean_centered_data) / (data_reshaped.size(0) - 1)
        
        # Compute eigenvalues and eigenvectors
        eigenvalues, eigenvectors = torch.linalg.eigh(cov_matrix)
        
        # Select the eigenvector corresponding to the largest eigenvalue
        principal_component = eigenvectors[:, -1]
        
        # Project the data onto the principal component
        reduced_data = torch.mm(mean_centered_data, principal_component.unsqueeze(1))
        
        # Reshape back to the original batch, H, W (with 1 channel)
        pca_images = reduced_data.view(B, H, W).unsqueeze(1)
        
        # normalize
        # Min-max normalization
        min_vals = torch.min(pca_images, dim=0).values
        max_vals = torch.max(pca_images, dim=0).values
        # Avoid division by zero in case max == min
        normalized_ims = (pca_images - min_vals) / (max_vals - min_vals + 1e-10)
        return normalized_ims 

    def forward(self, x, y_hat, y):
        # x: B, C, H, W
        # y_hat B, classes, H, W
        # partial cross entropy
        partial = self.partialCE(y_hat, y)
        # compute PCA
        # B, 1, H, W
        pca_imgs = self.compute_pca(x)

        # compute image tree
        img_mst_tree = self.mst(pca_imgs)

        # softmax y
        _, class_hat = torch.max(y_hat, dim=1)
        # Cls: B, H, W
        class_hat = class_hat.float()
        # y: B, H, W
        y = y.float()

        # psuedo label for global info
        soft_pseudo = self.global_apro(y_hat, pca_imgs, img_mst_tree, zeta_g=0.001)
        soft_pseudo = self.global_apro(soft_pseudo, y_hat, img_mst_tree, zeta_g=0.05)

        # unlabelled region only
        unlabelled_regions = (y == 0)

        # compute difference between psuedo labels and other
        loss_global_term = torch.abs(soft_pseudo-y) * unlabelled_regions
        # normalize the loss
        n_regions = unlabelled_regions.sum().clamp(min=1)
        loss_global = loss_global_term.sum() / n_regions

        # local term
        soft_pseudo = self.local_apro(pca_imgs, y)
        loss_local_term = torch.abs(y - soft_pseudo) * unlabelled_regions
        loss_local_term = loss_local_term.sum() / unlabelled_regions.sum()
        loss_local = loss_local_term
        return partial + loss_global + loss_local


class AproLightning(SparseSegLightning):
    def __init__(self,
                 net: Callable,
                 loss: Callable):
        super().__init__(
            net=net,
            loss=loss
        )

    def compute_loss(self, x, y_hat, target):
        return self.criterion(x, y_hat, target["mask"])
```


After implementing your changes, see above.

I have a couple of questions about input sizes which could be causing a problem.

1. Can the MST take in a 1 channel image?

2. You changed the `global_apro()` function to run twice, once with output logits (I think this should be output classes no? Doesn't work with logits, I changed to class values (class_hat (B, H, W) and the loss definitely seems to decrease but the output metrics are the same if you just the partial cross entropy.

Look forward to hearing your response.

If possible some pseudo code explaining like you do the README, but with point wise annotations would be great.

**Comment by CircleRadon** [comment]
Sorry for the delayed reply.
1. A 1-channel image is acceptable; additionally, images with more than 3 channels are also fine.
2. Apologies for any confusion earlier in the code – there may have been a mix-up between y and y_hat. The code has been corrected below, and if I have any further misunderstandings, please let me know.

By the way, the purpose of the APro loss is to guide the y_hat using optimized pseudo labels; there's no need to convert these to class values.
```python
   def forward(self, x, y_hat, y):
        # x: B, C, H, W
        # y_hat B, classes, H, W
        # partial cross entropy
        partial = self.partialCE(y_hat, y)
        # compute PCA
        # B, 1, H, W
        pca_imgs = self.compute_pca(x)

        # compute image tree
        # I think directly using x is also fine
        # img_mst_tree = self.mst(x)
        img_mst_tree = self.mst(pca_imgs)

        # y: B, H, W
        y = y.float()

        # psuedo label for global info
        # using low level feature
        soft_pseudo = self.global_apro(y_hat, pca_imgs, img_mst_tree, zeta_g=0.001)
        # using deep feature
        soft_pseudo = self.global_apro(soft_pseudo, y_hat, img_mst_tree, zeta_g=0.05)

        # unlabelled region only
        unlabelled_regions = (y == 0)

        # compute difference between generated psuedo labels and predicted one
        loss_global_term = torch.abs(soft_pseudo-y_hat) * unlabelled_regions
        # normalize the loss
        n_regions = unlabelled_regions.sum().clamp(min=1)
        loss_global = loss_global_term.sum() / n_regions

        # local term
        soft_pseudo = self.local_apro(pca_imgs, y_hat)
        loss_local_term = torch.abs(y_hat - soft_pseudo) * unlabelled_regions
        loss_local_term = loss_local_term.sum() / unlabelled_regions.sum()
        loss_local = loss_local_term
        return partial + loss_global + loss_local
```

**Comment by alfieroddan** [comment]
Thanks so much for your help! Will try again with this :).
